### ISSUE TITLE
Keras mixed precision API 50x slower than mixed precision graph rewrite

### ISSUE THREAD

**Issue by lgeiger** [tensorflow/tensorflow#41715]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.2.0, 2.3.0rc2, 2.4.0.dev2020072401
- Python version: 3.7.8
- CUDA/cuDNN version: 10.1 / 7.6.5.32
- GPU model and memory: NVIDIA V100

I currently use the [mixed precision graph rewrite](https://www.tensorflow.org/api_docs/python/tf/train/experimental/enable_mixed_precision_graph_rewrite) which will apply mixed precision optimizations to the entire TF graph. Since 2.1 there is a [new Keras API](https://www.tensorflow.org/guide/keras/mixed_precision) which replaces the graph rewrite and provides layerwise customization of the dtype policy. I tried switching to the new API and ran into the following issue.

**Describe the current behavior**

The Keras mixed precision API is 50 times slower than the legacy graph rewrite and the float32 baseline. When running the code example below on a NVIDIA V100 I get the following runtimes:

|  | float32 | Keras AMP | Keras AMP (float32 depthwise) | AMP graph rewrite
--- | --- | --- | --- | ---
time / step | 120ms  | **2000ms** | 38ms | 37ms
time / epoch | 6s | **90s** | 2s | 2s

**Describe the expected behavior**

Performance of the Keras mixed precision API should be on par with the graph rewrite and should outperform normal float32 training.

**Standalone code to reproduce the issue**
```python
import tensorflow as tf
import tensorflow_datasets as tfds

policy = tf.keras.mixed_precision.experimental.Policy("mixed_float16")
tf.keras.mixed_precision.experimental.set_policy(policy)


def preprocessing(data):
    return tf.cast(data["image"], tf.float32) / 255.0, data["label"]


dataset = (
    tfds.load("cifar10", split="train")
    .map(preprocessing, num_parallel_calls=tf.data.experimental.AUTOTUNE)
    .cache()
    .batch(1024)
    .prefetch(1)
)

model = tf.keras.models.Sequential(
    [
        tf.keras.layers.Conv2D(8, 3, padding="same", activation="relu", input_shape=(32, 32, 3)),
        tf.keras.layers.DepthwiseConv2D(3, depth_multiplier=8, padding="same", activation="relu"),
        tf.keras.layers.GlobalAveragePooling2D(),
        tf.keras.layers.Dense(10),
        tf.keras.layers.Activation("softmax", dtype="float32"),
    ]
)

model.compile(optimizer="adam", loss="sparse_categorical_crossentropy")

model.fit(dataset, epochs=3, callbacks=[tf.keras.callbacks.TensorBoard("logs")])
```

**Other info / logs**

This big slowdown is cause by the depthwise convolution in the model which takes up the bulk of the computation in the toy model mentioned above. When looking at the generated profile the issue it looks like Keras AMP runs the very slow
```cpp
DepthwiseConv2dBackpropFilterGPUKernelNHWC(DepthwiseArgs, Eigen::half const*, Eigen::half const*, Eigen::half*, int)
```
kernel whereas `float32` runs
```cpp
DepthwiseConv2dBackpropFilterGPUKernelNCHW(DepthwiseArgs, float const*, float const*, float*, int)
```
and the AMP graph rewrite executes
```cpp
DepthwiseConv2dBackpropFilterGPUKernelNHWC(DepthwiseArgs, float const*, float const*, float*, int)
```
It looks like the `float32` versions of the kernels are a lot faster, although I am not sure why the baseline uses `NCHW` instead of `NHWC`.

The underlying issue is definitely the unusable `float16` `DepthwiseConv2dBackpropFilter` kernel (see also #27780), although I think it would be good to automatically workaround this in Keras AMP or grappler as well since this makes Keras AMP not usable for networks with depthwise convolutions.

/cc @reedwm

**Comment by lgeiger** [comment]
For anyone stumbling across this issue as well, one can workaround the problem by explicitely setting `dtype="float32"` for `tf.keras.layers.DepthwiseConv2D`.

**Comment by dathudeptrai** [comment]
@lgeiger I also got this problem on both DepthWiseConv and GroupConv, seem mixed-precision not yet optimize for those layers. Also, I see in cuDNN 8.0 releases notes states that mixed-precision now support for GroupConv, I don't know if it also fixes DepthWiseConv. (https://docs.nvidia.com/deeplearning/sdk/cudnn-release-notes/rel_8.html)

**Comment by amahendrakar** [comment]
@lgeiger,
Is this still an issue? Could you please let us know if @dathudeptrai's workaround has fixed the issue. Thanks!

**Comment by lgeiger** [comment]
@amahendrakar yes, this is stil an issue. As mentioned in the issue description switching to an entirely different API is a valid workaround, but it doesn't fix the underlying issue with the Keras mixed precision API. It is the API that is recommended in most guides on the website so many people will likely run into this issue when trying to follow the guides.

**Comment by reedwm** [comment]
It looks like the backward filter op is slow with fp16, but not the forward or backward input op. 

I think a grappler pass is changing the data_format, as NHWC is typically faster for fp16 and NCHW is typically faster for fp32. Both data formats are extremely slow in fp16 and fast in fp32, so the data_format issue is a bit of a red herring.

The function which is slow in fp16 and fast in fp32 is [DepthwiseConv2dBackpropFilterGPUKernelNHWC](https://github.com/tensorflow/tensorflow/blob/1da0ebb453374dd3b1b30272cda11c27dd9dc276/tensorflow/core/kernels/depthwise_conv_op_gpu.h#L1030). I'm not sure why it's so slow in fp16 though. Maybe the atomics are slower in fp16? CC @chsigg can you fix this?

Also the comment at the top of the function makes me worried about the numeric stability:

https://github.com/tensorflow/tensorflow/blob/1da0ebb453374dd3b1b30272cda11c27dd9dc276/tensorflow/core/kernels/depthwise_conv_op_gpu.h#L1024-L1033

@nluehr doing accumulation in fp16 is dangerous, right? I think I'll have the Keras API force `DepthwiseConv2D` in float32 when cudnn is not used, both for performance and numeric stability



Here is a benchmark to verify the slowness:

```python
import tensorflow as tf
import time

def bench(dtype, data_format, func):
  if data_format == 'NHWC':
    x = tf.random.normal((32, 32, 32, 8), dtype=dtype)
    out_grad = tf.random.normal((32, 32, 32, 64), dtype=dtype)
  else:
    x = tf.random.normal((32, 8, 32, 32), dtype=dtype)
    out_grad = tf.random.normal((32, 64, 32, 32), dtype=dtype)
  f = tf.random.normal((3, 3, 8, 8), dtype=dtype)

  p = tf.constant(0.)

  def run():
    if func == 'forward':
      tf.nn.depthwise_conv2d(x, f, [1, 1, 1, 1], 'SAME',
                             data_format=data_format)
    elif func == 'backprop_filter':
      tf.nn.depthwise_conv2d_backprop_filter(
          x, f.shape, out_grad, [1, 1, 1, 1], 'SAME', data_format=data_format)
    else:
      assert func == 'backprop_input'
      tf.nn.depthwise_conv2d_backprop_input(
          x.shape, f, out_grad, [1, 1, 1, 1], 'SAME', data_format=data_format)

  # Warmup
  run()

  start = time.time()
  for _ in range(10):
    run()
  # Synchronize GPU by sending result of computation to CPU
  p = p + 1.
  p.numpy()

  end = time.time()
  print('time for %s %s %s: %s' % (dtype, data_format, func, end - start))

bench('float32', 'NHWC', 'forward')
bench('float32', 'NCHW', 'forward')
bench('float16', 'NHWC', 'forward')
bench('float16', 'NCHW', 'forward')
bench('float32', 'NHWC', 'backprop_input')
bench('float32', 'NCHW', 'backprop_input')
bench('float16', 'NHWC', 'backprop_input')
bench('float16', 'NCHW', 'backprop_input')
bench('float32', 'NHWC', 'backprop_filter')
bench('float32', 'NCHW', 'backprop_filter')
bench('float16', 'NHWC', 'backprop_filter')
bench('float16', 'NCHW', 'backprop_filter')
```

The output is

```
time for float32 NHWC forward: 0.002748727798461914
time for float32 NCHW forward: 0.0024683475494384766
time for float16 NHWC forward: 0.0024971961975097656
time for float16 NCHW forward: 0.0023937225341796875
time for float32 NHWC backprop_input: 0.0015697479248046875
time for float32 NCHW backprop_input: 0.0016202926635742188
time for float16 NHWC backprop_input: 0.0015521049499511719
time for float16 NCHW backprop_input: 0.0016052722930908203
time for float32 NHWC backprop_filter: 0.010046243667602539
time for float32 NCHW backprop_filter: 0.04735541343688965
time for float16 NHWC backprop_filter: 7.651637315750122
time for float16 NCHW backprop_filter: 4.234391450881958
```

**Comment by lgeiger** [comment]
@reedwm Thanks for investigating!

Looks like there is not much performance difference between fp16 and fp32 apart from the backprop filter kernel which is almost unusable in fp16, so I guess it won't hurt to disable fp16 for keras depthwise layers by default.

**Comment by lgeiger** [comment]
Thanks for the fix @reedwm, looks like this also closes #27780.

**Comment by reedwm** [comment]
Good catch, I should have gotten to that bug earlier. Thanks for the short, self-contained example to reproduce the issue!
